### DISCUSSION TITLE
Is RDF4J vulnerable to the new log4j exploit?

### DISCUSSION THREAD

**Discussion by hmottestad** [eclipse-rdf4j/rdf4j#3503] [Q&A] [answered]
I know that we don't use log4j, but we do use the log4j over slf4j adapter.

Is log4j used by any of our dependencies?

https://www.lunasec.io/docs/blog/log4j-zero-day/

**Comment by barthanssens** [comment]
By elasticsearch I think (might be compliance test only)... I've just seen a warning from dependabot

Best regards,

Bart

________________________________
From: hmottestad ***@***.***>
Sent: Friday, December 10, 2021 9:50
To: eclipse/rdf4j ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [eclipse/rdf4j] Is RDF4J vulnerable to the new log4j exploit (Discussion #3503)


I know that we don't use log4j, but we do use the log4j over slf4j adapter.

Is log4j used by any of our dependencies?

https://www.lunasec.io/docs/blog/log4j-zero-day/

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/eclipse/rdf4j/discussions/3503>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAOHUPJWVMWIWAEE7TI4HFTUQG5MNANCNFSM5JYSSKAQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by barthanssens** [comment]
Indeed, used in https://github.com/eclipse/rdf4j/blob/main/compliance/elasticsearch/pom.xml
I'll create a github ticket for it.

Bart
________________________________

**Comment by abrokenjester** [comment, marked as answer]
To clarify for users of RDF4J:

RDF4J as a framework is logger-agnostic (using [SLF4J](http://www.slf4j.org/)), so for RDF4J users this will not require a version update of their RDF4J dependencies. Reading the [CVE](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-44228), it is only the log4j-core library that is affected. The log4j-over-slf4j adapter is maintained by the SLF4J project and is not affected by this vulnerability.

In summary: 

- if you use RDF4J Server, Workbench, or Console, you do not need to update: these applications use Apache Commons Logging or Java Util logging rather than Log4J. 
- if you use RDF4J as a dependency in your own Java project, you do not need to update to a newer version of RDF4J. However you may need to update your log4j-core dependency if that is your [logger of choice](https://rdf4j.org/documentation/programming/setup/#logging-slf4j-initialization). 

The only places log4j-core is referenced in our code base is as test-only dependencies for the elasticsearch, solr, and lucene modules. We are fixing those (thanks for the quick reaction @barthanssens !). However, those test dependencies are not distributed as releases and will not affect normal usage of RDF4J.

**Comment by hmottestad** [comment]
Seems that log4j is in use in the SolrSail at least.

Command: `mvn dependency:list -Dsort=true -DincludeGroupIds=org.apache.logging.log4j -DincludeScope=compile `

```

----------------------< org.eclipse.rdf4j:rdf4j >-----------------------
-----------------< org.eclipse.rdf4j:rdf4j-sail-solr >------------------
Building RDF4J: Solr Sail Index 4.0.0-SNAPSHOT                  [75/112]
--------------------------------[ jar ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-sail-solr ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-1.2-api:jar:2.11.2:compile
   org.apache.logging.log4j:log4j-api:jar:2.11.2:compile
   org.apache.logging.log4j:log4j-core:jar:2.11.2:compile
   org.apache.logging.log4j:log4j-slf4j-impl:jar:2.11.2:compile
   org.apache.logging.log4j:log4j-web:jar:2.11.2:compile


-------------< org.eclipse.rdf4j:rdf4j-sail-elasticsearch >-------------
Building RDF4J: Elastic Search Sail Index 4.0.0-SNAPSHOT        [76/112]
--------------------------------[ jar ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-sail-elasticsearch ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-api:jar:2.11.1:compile


----------< org.eclipse.rdf4j:rdf4j-sail-elasticsearch-store >----------
Building RDF4J: Elasticsearch Store 4.0.0-SNAPSHOT              [78/112]
--------------------------------[ jar ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-sail-elasticsearch-store ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-api:jar:2.11.1:compile


-------------< org.eclipse.rdf4j:rdf4j-spring-components >--------------
Building RDF4J: Spring components 4.0.0-SNAPSHOT                [92/112]
--------------------------------[ pom ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-spring-components ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
   org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile


-----------< org.eclipse.rdf4j:rdf4j-spring-boot-sparql-web >-----------
Building RDF4J: Spring boot component for a HTTP sparql server 4.0.0-SNAPSHOT [93/112]
--------------------------------[ jar ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-spring-boot-sparql-web ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
   org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile


-------------------< org.eclipse.rdf4j:rdf4j-spring >-------------------
Building RDF4J: Spring 4.0.0-SNAPSHOT                           [94/112]
--------------------------------[ jar ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-spring ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
   org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile


----------------< org.eclipse.rdf4j:rdf4j-spring-demo >-----------------
Building RDF4J: Spring Demo 4.0.0-SNAPSHOT                      [95/112]
--------------------------------[ jar ]---------------------------------

--- maven-dependency-plugin:2.8:list (default-cli) @ rdf4j-spring-demo ---

The following files have been resolved:
   org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
   org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile



```

**Reply by hmottestad** [reply]
As far as I understand the vulnerability is with the `core` dependency.

**Reply by abrokenjester** [reply]
Ok if that is indeed the case, it is a mistake (regardless of the exploit): we should _never_ include a logger implementation in the library modules, only the apis and ...-to-slf4j adapters.

**Reply by abrokenjester** [reply]
Can you do a scan of the sdk zip / tgz files and see if the log4j-core lib is included there? If it is, we will need to do a patch release after all.

**Reply by barthanssens** [reply]
Hmz, can't find log4j-core in the 3.7.x / 4.0.0-M1 SDKs  lib directory,  nor in the war WEB-INF's or OSGI... but please double/triple-check :-)

(nor a log4j-over-slf4j adapter for that matter, so I'm wondering if logging without adding additional dependencies actually works for Solr/ES sails, we do include logback-core - which isn't affected)

**Reply by abrokenjester** [reply]
This log4j-core dependency comes transitively through solr-core. Luckily, solr-core itself is an optional dependency, which means we don't distribute it. This is why it doesn't show up in the SDK zip, despite being a compile-time dependency. 

So we're good, with respect to not needing a patch release. We do need to fix the pom in sail-solr, and make sure log4j-core is explicitly excluded.

**Reply by abrokenjester** [reply]
Minor followup: it appears the fix in log4j-core 2.15.0 was incomplete, and another patch release has been done: 2.16.0 (see also dependabot PR at https://github.com/eclipse/rdf4j/pull/3513). I've reopened #3504 to address this (and we can tackle the exclusion in sail-solr while we're at it, hopefully).

**Reply by barthanssens** [reply]
To a lesser extent logback-core might be vulnerable (one has to have write access to the config file), https://jira.qos.ch/browse/LOGBACK-1591
Opened #3516

**Reply by hmottestad** [reply]
The logback one is fairly low severity. You need both write access to the config file and access to force a reload of the config file into logback. 

With log4j the attacker just needed to trigger that user input was logged.
